Chess/checkvalidator.cpp

```diff
@@ -15,17 +39,38 @@ void CheckValidator::isCheck(Color player)
             piece=pieces[i];
             break;
         }
-    Color enemyColor;
-    if(player==_WHITE)
-        enemyColor=_BLACK;
-    else enemyColor=_WHITE;
-    std::vector<Piece*>enemy = board->getPieces(enemyColor);
-    size=enemy.size();
+    if(CheckValidator::isCheckPosition(piece->getRow(),piece->getColumn(),piece->getColor()))
+        std::cout<<"Sah, muta regele "<<piece->getColor()<<'\n';
     ///Need every valid move
     ///To be done as the last task
 }
 
 void CheckValidator::isCheckMate(Color playerToMove)
 {
-    ///Same as last task
+    Board* board = Board::getInstance();
+    std::vector<Piece*>pieces = board->getPieces(playerToMove);
+    Piece* piece;
+    int size=pieces.size();
+    for(int i=0;i<size;i++)
+        if(pieces[i]->getType()==KING)
+        {
+            piece=pieces[i];
+            break;
+        }
+    std::vector<MoveBy>validPositions=piece->getValidPositions();
+    int cnt=validPositions.size() + 1;
+    if(isCheckPosition(piece->getRow(), piece->getColumn(), piece->getColor()))
+        cnt--;
+
+    for(auto moves:validPositions)
+    {
+        int row=moves.getDiffRow()+piece->getRow();
+        int column=moves.getDiffColumn()+piece->getColumn();
+        if(isCheckPosition(row,column,piece->getColor()))
+            cnt--;
+    }
+    if(cnt==0)
+        std::cout<<"Checkmate!! "<<piece->getColor()<<" a pierdut!";
+    else
+        std::cout<<"Nu e checkmate!!";
```

Comment:
Incearca sa spargi functia in mai multe functii, mai mici

Reply:
https://medium.com/oril/10-tips-to-keep-your-code-more-clean-2fa9aafea1cf
Sfatul 5